presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsTracker.java

```diff
@@ -0,0 +1,140 @@
+import com.facebook.presto.spi.statistics.Estimate;
+import com.facebook.presto.spi.statistics.HistoricalPlanStatistics;
+import com.facebook.presto.spi.statistics.HistoryBasedPlanStatisticsProvider;
+import com.facebook.presto.spi.statistics.PlanStatistics;
+import com.facebook.presto.sql.planner.PlanHasher;
+import com.facebook.presto.sql.planner.planPrinter.PlanNodeStats;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.SystemSessionProperties.useHistoryBasedPlanStatisticsEnabled;
+import static com.facebook.presto.common.plan.PlanCanonicalizationStrategy.historyBasedPlanCanonicalizationStrategyList;
+import static com.facebook.presto.spi.resourceGroups.QueryType.INSERT;
+import static com.facebook.presto.spi.resourceGroups.QueryType.SELECT;
+import static com.facebook.presto.sql.planner.planPrinter.PlanNodeStatsSummarizer.aggregateStageStats;
+import static com.google.common.graph.Traverser.forTree;
+import static java.util.Objects.requireNonNull;
+
+public class HistoryBasedPlanStatisticsTracker
+{
+    private static final Logger LOG = Logger.get(HistoryBasedPlanStatisticsTracker.class);
+    private static final Set<QueryType> ALLOWED_QUERY_TYPES = ImmutableSet.of(SELECT, INSERT);
+
+    private final Supplier<HistoryBasedPlanStatisticsProvider> historyBasedPlanStatisticsProvider;
+    private final SessionPropertyManager sessionPropertyManager;
+    private final PlanHasher planHasher;
+
+    public HistoryBasedPlanStatisticsTracker(
+            Supplier<HistoryBasedPlanStatisticsProvider> historyBasedPlanStatisticsProvider,
+            SessionPropertyManager sessionPropertyManager,
+            PlanHasher planHasher)
+    {
+        this.historyBasedPlanStatisticsProvider = requireNonNull(historyBasedPlanStatisticsProvider, "historyBasedPlanStatisticsProvider is null");
+        this.sessionPropertyManager = requireNonNull(sessionPropertyManager, "sessionPropertyManager is null");
+        this.planHasher = requireNonNull(planHasher, "planHasher is null");
+    }
+
+    public void trackStatistics(QueryExecution queryExecution)
+    {
+        queryExecution.addFinalQueryInfoListener(queryInfo -> {
+            trackStatistics(queryInfo);
+        });
+    }
+
+    @VisibleForTesting
+    public HistoryBasedPlanStatisticsProvider getHistoryBasedPlanStatisticsProvider()
+    {
+        return historyBasedPlanStatisticsProvider.get();
+    }
+
+    private void trackStatistics(QueryInfo queryInfo)
+    {
+        if (!useHistoryBasedPlanStatisticsEnabled(queryInfo.getSession().toSession(sessionPropertyManager))) {
+            return;
+        }
+
+        // Only update statistics for successful queries
```

Comment:
I know we discussed this, but I don't remember what the conclusion was.  What was the solution for being able to optimize based on query failures?

Reply:
We can do it in this endpoint, but more work needs to be done in identifying what stats can be trusted, and what are not as the query did die in between and we have incomplete stats. We can probably store some hints on join nodes